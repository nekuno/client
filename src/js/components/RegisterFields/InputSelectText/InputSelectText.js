import PropTypes from 'prop-types';
import React, { Component } from 'react';
import Input from '../../ui/Input/Input.js';
import Chip from '../../ui/Chip/Chip.js';
import styles from './InputSelectText.scss';

export default class InputSelectText extends Component {

    static propTypes = {
        placeholder    : PropTypes.string,
        options        : PropTypes.array,
        selectedLabel  : PropTypes.string,
        chipsColor     : PropTypes.oneOf(['purple', 'blue', 'pink', 'green']),
        onChangeHandler: PropTypes.func,
        onClickHandler : PropTypes.func,
        selectedValues : PropTypes.array,
    };

    constructor(props) {
        super(props);

        this.handleChange = this.handleChange.bind(this);
        this.handleClick = this.handleClick.bind(this);

        let selected = [];

        if (props.selectedValues) {
            selected = props.selectedValues;
        }

        this.state = {
            selected: selected,
            suggested: []
        }
    }

    handleChange(text) {
        const {options} = this.props;
        const {selected} = this.state;

        this.setState({suggested: options.filter(
            option => text
                && option.text.toLowerCase().indexOf(text.toLowerCase()) !== -1
                && !selected.some(item => item.id === option.id)
        )});

        if (this.props.onChangeHandler) {
            this.props.onChangeHandler(text);
        }
    }

    handleClick(id) {
        const {options} = this.props;
        const {selected} = this.state;
        let newSelected = selected.slice(0);
        const index = selected.findIndex(option => option.id === id);
        console.log(newSelected);

        if (index !== -1) {
            newSelected.splice(index, 1);
        } else {
            console.log(options.find(option => option.id === id));
            const option = options.find(option => option.id === id);
            console.log(option);
            console.log(newSelected);
            const newOption = {
                value: option.id,
                image: null,
            };
            // newSelected.push(options.find(option => option.id === id));
            newSelected.push(newOption);
        }
        this.setState({selected: newSelected, suggested: []});

        this.refs["input"].clearValue();

        if (this.props.onClickHandler) {
            this.props.onClickHandler(newSelected.map(option => option.value));
        }
    }

    render() {
        const {placeholder, options, selectedLabel, chipsColor, selectedValues} = this.props;
        const {selected, suggested} = this.state;

        return (
            <div className={styles.inputSelectText}>
                <Input ref="input" placeholder={placeholder} searchIcon={true} size={'small'} onChange={this.handleChange} doNotScroll={true}/>

                {suggested.map((item, index) =>
                    <div key={index} className={styles.suggestedChip}>
                        <Chip onClickHandler={this.handleClick}
                              text={item.text}
                              value={item.id}
                              color={chipsColor}
                        />
                    </div>
                )}
                {selected.length > 0 ?
                    <div className={styles.selectedLabel + " small"}>{selectedLabel}</div> : null
                }

                {options.length > 0 ?
                    selected.map((item, index) =>
                        <div key={index} className={styles.selectedChip}>
                            <Chip onClickHandler={this.handleClick}
                                  text={options.find(x => x.id === item.value).text}
                                  value={options.find(x => x.id === item.value).id}
                                  color={chipsColor}
                                  selected={true}
                            />
                        </div>
                    )
                    : null
                }
            </div>
        );
    }
}