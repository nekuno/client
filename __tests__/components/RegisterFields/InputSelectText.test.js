import React from 'react';
import InputSelectText from '../../../src/js/components/RegisterFields/InputSelectText/InputSelectText.js';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('Test InputSelectText component', () => {
    const mockFn = jest.fn();
    const options = [
        {
            id: "lorem",
            text: "Lorem"
        },
        {
            id: "ipsum",
            text: "Ipsum"
        },
        {
            id: "lorem-ipsum",
            text: "Lorem ipsum"
        },
        {
            id: "sed-ut-perspiciatis",
            text: "Sed ut perspiciatis"
        },
        {
            id: "perspiciatis",
            text: "perspiciatis"
        },
    ];

    const inputSelectText = shallow(<InputSelectText onClickHandler={mockFn} placeholder={"Foo"} options={options}/>);

    it('should be defined', () => {
        expect(InputSelectText).toBeDefined();
    });
    it('should render correctly', () => {
        expect(inputSelectText).toMatchSnapshot();
    });
});
