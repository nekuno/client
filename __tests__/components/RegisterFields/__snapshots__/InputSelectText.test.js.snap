// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test InputSelectText component should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <InputSelectText
    onClickHandler={[MockFunction]}
    options={
      Array [
        Object {
          "id": "lorem",
          "text": "Lorem",
        },
        Object {
          "id": "ipsum",
          "text": "Ipsum",
        },
        Object {
          "id": "lorem-ipsum",
          "text": "Lorem ipsum",
        },
        Object {
          "id": "sed-ut-perspiciatis",
          "text": "Sed ut perspiciatis",
        },
        Object {
          "id": "perspiciatis",
          "text": "perspiciatis",
        },
      ]
    }
    placeholder="Foo"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Input
          doNotScroll={true}
          onChange={[Function]}
          placeholder="Foo"
          searchIcon={true}
          size="small"
        />,
        Array [],
        null,
        Array [],
      ],
      "className": "inputSelectText",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "doNotScroll": true,
          "onChange": [Function],
          "placeholder": "Foo",
          "searchIcon": true,
          "size": "small",
        },
        "ref": "input",
        "rendered": null,
        "type": [Function],
      },
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Input
            doNotScroll={true}
            onChange={[Function]}
            placeholder="Foo"
            searchIcon={true}
            size="small"
          />,
          Array [],
          null,
          Array [],
        ],
        "className": "inputSelectText",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "doNotScroll": true,
            "onChange": [Function],
            "placeholder": "Foo",
            "searchIcon": true,
            "size": "small",
          },
          "ref": "input",
          "rendered": null,
          "type": [Function],
        },
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
